src/components/input-time/input-time.tsx

```diff
@@ -0,0 +1,872 @@
+
+    this.userChangedValue = context === "user";
+
+    const valueChanged = this.userChangedValue && formattedNewValue !== oldValue;
+
+    this.value = formattedNewValue;
+
+    if (context !== "user") {
+      this.hour = hour;
+      this.localizedHour = localizedHour;
+      this.localizedHourSuffix = getLocaleHourSuffix(this.locale);
+
+      this.minute = minute;
+      this.localizedMinute = localizedMinute;
+      this.localizedMinuteSuffix = getLocaleMinuteSuffix(this.locale);
+
+      this.second = second;
+      this.localizedSecond = localizedSecond;
+      this.localizedSecondSuffix = getLocaleSecondSuffix(this.locale);
+
+      this.meridiem = getMeridiem(this.hour);
+      this.localizedMeridiem = localizedMeridiem;
+      const formatParts = getTimeParts({ value: newValue, locale: this.locale });
+      this.meridiemOrder = this.getMeridiemOrder(formatParts);
+    }
+
+    if (valueChanged) {
+      this.previousValue = oldValue;
+
+      const inputEvent = this.calciteInputTimeInput.emit();
+
+      if (inputEvent.defaultPrevented) {
+        this.userChangedValue = false;
+        this.value = oldValue;
+        this.previousValue = oldValue;
+      } else {
+        this.previousValue = formattedNewValue;
+      }
+    }
+  };
+
+  private setValuePart = (
+    key: "hour" | "minute" | "second" | "meridiem",
+    value: number | string | Meridiem
+  ): void => {
+    const { locale } = this;
+    if (key === "meridiem") {
+      this.meridiem = value as Meridiem;
+      if (isValidNumber(this.hour)) {
+        const hourAsNumber = parseInt(this.hour);
+        switch (value) {
+          case "AM":
+            if (hourAsNumber >= 12) {
+              this.hour = formatTimePart(hourAsNumber - 12);
+            }
+            break;
+          case "PM":
+            if (hourAsNumber < 12) {
+              this.hour = formatTimePart(hourAsNumber + 12);
+            }
+            break;
+        }
+        this.localizedHour = localizeTimePart({
+          value: this.hour,
+          part: "hour",
+          locale
+        });
+      }
+    } else {
+      this[key] = typeof value === "number" ? formatTimePart(value) : value;
+      this[`localized${capitalize(key)}`] = localizeTimePart({
+        value: this[key],
+        part: key,
+        locale
+      });
+    }
+    if (this.hour && this.minute) {
+      const showSeconds = this.second && this.showSecond;
+      this.setValue({
+        newValue: `${this.hour}:${this.minute}:${showSeconds ? this.second : "00"}`
+      });
+    } else {
+      if (this.value) {
+        this.setValue({ newValue: null });
+      }
+    }
+    this.localizedMeridiem = this.value
+      ? localizeTimeStringToParts({ value: this.value, locale })?.localizedMeridiem || null
+      : localizeTimePart({ value: this.meridiem, part: "meridiem", locale });
+  };
+
+  private timePartFocusHandler = (event: FocusEvent): void => {
+    this.activeEl = event.currentTarget as HTMLSpanElement;
+  };
```

Comment:
Oh man, there's lot of if/else and case/switch statements in all the methods and handlers above lol 😵
It would be nice to refactor this at some point to make it more readable, maybe create some functional utils or something.

Reply:
agreed. I'd like to start pulling as many private functions out into their own single testable files in a ./[component]/functions/[fileName]` pattern. It will let us better test our components and maybe recognize more functional patterns we already have.